libcxx/test/std/utilities/function.objects/func.wrap/func.wrap.move/assignment/move.pass.cpp

```diff
@@ -0,0 +1,102 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// UNSUPPORTED: c++03, c++11, c++14, c++17, c++20
+
+#include <cassert>
+#include <functional>
+
+#include "test_macros.h"
+#include "../common.h"
+
+template <class T>
+void test() {
+  {
+    std::move_only_function<T> f = &call_func;
+    std::move_only_function<T> f2;
+    f2 = std::move(f);
+    assert(f2);
```

Comment:
Throughout (as in my branch): it would be useful to check at the same time that `assert(!f)` after it's been moved-from.

Reply:
I don't think we can do that. An implementation can say that it's engaged after it's been moved from. For example, in our case we could have special moved-from functions that we assign to `__call_` and `__destroy_`.

Reply:
Ah, you're correct that vendors are _permitted_ to leave a moved-from `move_only_function` in an engaged state: https://eel.is/c++draft/func.wrap.move#ctor-3
But we can certainly use `LIBCPP_ASSERT(!f)` here, and I recommend doing so.

(Also, no sane vendor will ever actually _take advantage of_ that permission. A default-constructed `move_only_function` must be disengaged; only a malicious implementation would gratuitously make the moved-from state _different_ from that default-constructed state. I'm confident that you could have written `assert(!f)` and nobody would ever have noticed that the test was overconstrained.)